### Numpy 라이브러리

- 빅데이터 수학적, 과학적 계산을 위해 만들어진 라이브러리
- 복문 없이 전체 데이터의 연산이 가능한 표준 수학 함수들을 제공
- 배열: 같은 자료형의 값들이 연속적인 형태로 구성된 자료 구조로 대량의 데이터를 그룹으로 관리하고 계산하기 위해 사용!
  
array : 배열로 list와 비슷하지만 연산을 위해 동이한 자료형만을 요소로 받는다.

In [1]:
# !pip install numpy

In [4]:
import numpy as np

### 1차원 배열

In [95]:
list1= [1,2,3,4,5]
print(type(list1))

# list to array
arr1 = np.array(list1)
print(type(arr1))

# array는 자료형이 같아야해서 1,2,3이 str으로 변한다
arr0=np.array([1,2,3, "야호!"])
print(type(arr0[1]))

# array to list
print(type(arr0.tolist()))

print('-'*115)

# 1차원 배열 생성 (n) : 0~n-1
arr_ar=np.arange(10)
print(arr_ar)

print('-'*115)

# 1차원 배열 생성 후 reshape 생성 갯수랑 행x열 이 같아야한다.
arr_ar2=np.arange(10).reshape(5,2)
print(arr_ar2)

<class 'list'>
<class 'numpy.ndarray'>
<class 'numpy.str_'>
<class 'list'>
-------------------------------------------------------------------------------------------------------------------
[0 1 2 3 4 5 6 7 8 9]
-------------------------------------------------------------------------------------------------------------------
[[0 1]
 [2 3]
 [4 5]
 [6 7]
 [8 9]]


### 2차원 배열

In [17]:
arr2 = np.array([[1,2,3,], [4, 5, 6]])

# 배열의 형태 확인(행, 열)
print("배열의 형태 :", arr2.shape)

# 배열의 요소 전체 개수
print("배열의 전체 요소개수 :", arr2.size)

# 배열의 데이터 타입
print("배열의 데이터 타입 :", arr2.dtype)   #int64의 의미는 2**64 = -2**32 부터 +2**32-1까지 
                    #32는 bit(0,1) 2의 32제곱 의미
                    #int8 -> 1비트당 2개의 숫자를 표현할 수 있음 2**8 = 256개 표현 가능 (-128~127)
                    #int32 -> -2,147,483,648 ~ 2,147,483,647
                    
# 배열의 차원 확인
print("배열의 차원 :", arr2.ndim)

배열의 형태 : (2, 3)
배열의 전체 요소개수 : 6
배열의 데이터 타입 : int64
배열의 차원 : 2


In [78]:
# 0으로 배열 생성(차원, 행, 열) // 두개만 적으면 차원 빼고 (행,열)로 인식한다.
arr_zeros=np.zeros((3,4))
print(arr_zeros)

print('-'*115)

# 1로 배열 생성(차원, 행, 열)
arr_ones=np.ones((2,4,3))
print(arr_ones)

print('-'*115)

# 특정값으로 배열 생성하기 ((차원, 행, 열), 특정값)
arr_full = np.full((3,4,3),2)
print(arr_full)

print('-'*115)

# 범위로 배열 생성 (시작값, 미만값, 증감폭)
arr_range=np.arange(1,10)
print(arr_range)

[[0. 0. 0. 0.]
 [0. 0. 0. 0.]
 [0. 0. 0. 0.]]
-------------------------------------------------------------------------------------------------------------------
[[[1. 1. 1.]
  [1. 1. 1.]
  [1. 1. 1.]
  [1. 1. 1.]]

 [[1. 1. 1.]
  [1. 1. 1.]
  [1. 1. 1.]
  [1. 1. 1.]]]
-------------------------------------------------------------------------------------------------------------------
[[[2 2 2]
  [2 2 2]
  [2 2 2]
  [2 2 2]]

 [[2 2 2]
  [2 2 2]
  [2 2 2]
  [2 2 2]]

 [[2 2 2]
  [2 2 2]
  [2 2 2]
  [2 2 2]]]
-------------------------------------------------------------------------------------------------------------------
[1 2 3 4 5 6 7 8 9]


In [93]:
# 배열 행렬 전치 
arr3=np.zeros((2,4))
print(arr3)
arr_t=arr3.T  # 전치를 위해서는 대입 연산자 사용해야함
print(arr_t)

print('-'*115)

# 배열의 행과 열 재배치
arr3.reshape(2,4)
print(arr3)

print('-'*115)

# 자료형태 변경하기 np.array(listName, dtpye=np.자료형)
arr_ch=np.array(arr3, dtype=np.str)
print(type(arr_ch[0][0]))


[[0. 0. 0. 0.]
 [0. 0. 0. 0.]]
[[0. 0.]
 [0. 0.]
 [0. 0.]
 [0. 0.]]
-------------------------------------------------------------------------------------------------------------------
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]]
-------------------------------------------------------------------------------------------------------------------
<class 'numpy.str_'>


/var/folders/d0/hsqd9x956qgdv8tn9cts8vmc0000gn/T/ipykernel_59245/1644775871.py:16: DeprecationWarning: `np.str` is a deprecated alias for the builtin `str`. To silence this warning, use `str` by itself. Doing this will not modify any behavior and is safe. If you specifically wanted the numpy scalar type, use `np.str_` here.
Deprecated in NumPy 1.20; for more details and guidance: https://numpy.org/devdocs/release/1.20.0-notes.html#deprecations
  arr_ch=np.array(arr3, dtype=np.str)


In [92]:
# 랜덤 난수(0~1) 생성하기 (행,열)
arr_rand=np.random.rand(2,3)
print(arr_rand)

print('-'*115)

# 랜덤 정수 1개 생성하기 (미만값)
arr_randint0=np.random.randint(10)
print(arr_randint0)

print('-'*115)

# 1차원 랜덤 정수 배열 생성하기 (미만값, size=개수)
arr_randint1=np.random.randint(10, size=4)
print(arr_randint1)

print('-'*115)

# 2차원 랜덤 정수 배열 생성하기 (미만값, size=(행,열))
arr_randint2=np.random.randint(10, size=(3,2))
print(arr_randint2)

[[0.64238805 0.32247669 0.71007473]
 [0.94601337 0.83554535 0.41529905]]
-------------------------------------------------------------------------------------------------------------------
6
-------------------------------------------------------------------------------------------------------------------
[5 2 9 8]
-------------------------------------------------------------------------------------------------------------------
[[4 9]
 [6 9]
 [7 3]]


### 배열의 연산

In [77]:
list4 = [[1,2],
          [3,4]]
print(list4+list4)

[[1, 2], [3, 4], [1, 2], [3, 4]]


In [113]:
# 배열 사칙연산 (덧셈 곱셈 뺄셈 나눗셈 다 됨)
arr4 = np.array([[1,2],
                 [3,4]])
print(arr4+arr4)

# 합계 (배열 요소 다 합침)
print('합계', np.sum(arr4))
print('합계', arr4.sum())

# 평균
print('평균', np.mean(arr4))
print('평균', arr4.mean())

# 중앙값
print('중앙값', np.median(arr4))

# 제곱근
print('제곱근') 
print(np.sqrt(arr4))

# 절대값
print('절대값') 
print(np.abs(arr4))

[[2 4]
 [6 8]]
합계 10
합계 10
평균 2.5
평균 2.5
중앙값 2.5
제곱근
[[1.         1.41421356]
 [1.73205081 2.        ]]
절대값
[[1 2]
 [3 4]]


### 인덱싱/슬라이싱

- 인덱싱 : 배열 중 하나의 인덱스값에 있는 데이터를 가져오는 것
- 슬라이싱 : 배열의 범위를 지정하여 여러개의 데이터를 가져오는 것

In [96]:
# 인덱싱
arr5 = np.array([[1,2,3],
                 [4,5,6]])
print(arr5[0][1])

print('-'*115)

# 슬라이싱
arr6 = np.arange(10)
print(arr6[3:8])

print('-'*115)

# 2차원 데이터 슬라이싱(행시작:행끝, 열시작,열끝)
arr7=np.arange(50).reshape(5,10)
print(arr7)

print('-'*115)

print(arr7[0:2, 0:2])

2
-------------------------------------------------------------------------------------------------------------------
[3 4 5 6 7]
-------------------------------------------------------------------------------------------------------------------
[[ 0  1  2  3  4  5  6  7  8  9]
 [10 11 12 13 14 15 16 17 18 19]
 [20 21 22 23 24 25 26 27 28 29]
 [30 31 32 33 34 35 36 37 38 39]
 [40 41 42 43 44 45 46 47 48 49]]
-------------------------------------------------------------------------------------------------------------------
[[ 0  1]
 [10 11]]


In [91]:
arr19 = np.arange(18).reshape(3,6)
arr19

array([[ 0,  1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10, 11],
       [12, 13, 14, 15, 16, 17]])

### 불리언 인덱싱

- 특정 조건에 맞으면 True, 아니면 False 값으로 배열을 만들어 조건에 맞는 값에 접근하는 방법

In [104]:
arr10 = np.array(np.random.randint(50,100, size=8))
print(arr10)

#boolean(불리언) : True(참) 또는 False(거짓)으로 출력되는 자료형태
print(arr10 > 60)

# 불리언 인덱싱
arr10[arr10 > 60]

[79 94 72 64 79 77 54 56]
[ True  True  True  True  True  True False False]


array([79, 94, 72, 64, 79, 77])

## 예제) BMI지수 및 과체중 BMI값 확인

#### 데이터 로드

In [115]:
#delimiter(구분자) : 파일에서 데이터를 구분해주는 문자 지정
data = np.loadtxt('rowdata/height_weight.txt', delimiter=",")
print(data)  # 키는 cm / 몸무게는 kg

[[175.2 180.3 175.  169.2 185.2 188.  177.6 178.2 177.  179. ]
 [ 65.6  88.   79.2  69.3  55.   71.2  73.   68.9  74.   82. ]]


#### 데이터의 속성확인
- 배열 형태
- 배열 차원수
- 배열 전체요소 개수

In [116]:
print("배열 형태:",data.shape)
print("배열 차원수:",data.ndim)
print("배열 전체요소 개수:",data.size)

배열 형태: (2, 10)
배열 차원수: 2
배열 전체요소 개수: 20


#### 데이터에서 키와 몸무게 데이터들을 각각 변수에 담기

In [117]:
#키
height = data[0]
#몸무게
weight = data[1]

print(height)
print(weight)

[175.2 180.3 175.  169.2 185.2 188.  177.6 178.2 177.  179. ]
[65.6 88.  79.2 69.3 55.  71.2 73.  68.9 74.  82. ]


In [118]:
bmi=weight/((height/100)**2)
print(bmi)

[21.37153104 27.07018468 25.86122449 24.20652885 16.03543423 20.14486193
 23.14392095 21.69720651 23.62028791 25.59220998]


#### BMI지수가 23이상 25이하인 사람은 몇 명일까요?
- 불리언 인덱싱에서는 and -> & , or -> |, not->~를 쓴다

In [119]:
bmi[(bmi >= 23) & (bmi <= 25)]

array([24.20652885, 23.14392095, 23.62028791])

In [120]:
len(bmi[(bmi >= 23) & (bmi <= 25)])

3

## 예제) 영화 리뷰 평점 분석하기!

#### 데이터 로드

In [121]:
movie = np.loadtxt('rowdata/ratings.dat', delimiter='::', dtype=np.int64)
movie
#사용자 번호(user_id), 영화번호(item_id), 평점(ratings), 입력시간(time)

array([[        1,      1193,         5, 978300760],
       [        1,       661,         3, 978302109],
       [        1,       914,         3, 978301968],
       ...,
       [     6040,       562,         5, 956704746],
       [     6040,      1096,         4, 956715648],
       [     6040,      1097,         4, 956715569]])

In [122]:
print('배열 형태', movie.shape)
print('배열 차원수', movie.ndim)
print('배열 전체요소 개수', movie.size)

배열 형태 (1000209, 4)
배열 차원수 2
배열 전체요소 개수 4000836


#### 데이터에 접근하기

In [123]:
ratings=movie[:, 2]

In [124]:
#mean함수 사용하기
np.mean(ratings)

3.581564453029317

#### 1번 사용자가 매긴 평점의 평균 구하기
- 사용자 번호의 유일한 값(유니크 값)확인
- 사용자 번호의 유니크 값 개수 확인
- 1번 사용자의 데이터에 접근(불리언 인덱싱)
- 1번 사용자가 매긴 평점들의 평균 구하기

In [128]:
#평점을 매긴 사용자가 몇 명인지 알아보자!
user_id=np.unique(movie[:,0])
print(user_id)

[   1    2    3 ... 6038 6039 6040]


In [126]:
user_id.size

6040

In [127]:
#사용자의 번호가 1번인 데이터만 출력
a=movie[ movie[:,0] == 1 ]
print(a)

[[        1      1193         5 978300760]
 [        1       661         3 978302109]
 [        1       914         3 978301968]
 [        1      3408         4 978300275]
 [        1      2355         5 978824291]
 [        1      1197         3 978302268]
 [        1      1287         5 978302039]
 [        1      2804         5 978300719]
 [        1       594         4 978302268]
 [        1       919         4 978301368]
 [        1       595         5 978824268]
 [        1       938         4 978301752]
 [        1      2398         4 978302281]
 [        1      2918         4 978302124]
 [        1      1035         5 978301753]
 [        1      2791         4 978302188]
 [        1      2687         3 978824268]
 [        1      2018         4 978301777]
 [        1      3105         5 978301713]
 [        1      2797         4 978302039]
 [        1      2321         3 978302205]
 [        1       720         3 978300760]
 [        1      1270         5 978300055]
 [        1

In [129]:
#1번 사용자가 매긴 평점들
array = a[:,2]
print(array)

[5 3 3 4 5 3 5 5 4 4 5 4 4 4 5 4 3 4 5 4 3 3 5 5 3 5 4 4 4 3 4 4 4 4 4 4 5
 5 4 5 5 5 4 4 4 5 5 4 5 4 4 4 4]


In [130]:
np.mean(array)

4.188679245283019

- 1번 사용자는 전체 영화의 평점 평균보다 후하게 평점을 줬구나~!

#### 각 사용자가 매긴 평점들의 평균 구하기

In [96]:
movie[ movie[:,0]==i][:,2].mean()

3.7131782945736433

In [97]:
user_mean_list = []

for i in user_id:
    # 1번 ~6040번 사용자가 매긴 평점들의 평균을 각각 구해주는 코드
    user_mean = movie[ movie[:,0]==i][:,2].mean()
    # 빈 리스트에 사용자 번호와 그 사람이 매긴 평점의 평균을 리스트로 저장
    user_mean_list.append([i,user_mean])
user_mean_list

[[1, 4.188679245283019],
 [2, 3.7131782945736433],
 [3, 3.9019607843137254],
 [4, 4.190476190476191],
 [5, 3.1464646464646466],
 [6, 3.9014084507042255],
 [7, 4.32258064516129],
 [8, 3.884892086330935],
 [9, 3.7358490566037736],
 [10, 4.114713216957606],
 [11, 3.2773722627737225],
 [12, 3.8260869565217392],
 [13, 3.388888888888889],
 [14, 3.32],
 [15, 3.3233830845771144],
 [16, 3.0285714285714285],
 [17, 4.075829383886256],
 [18, 3.6491803278688524],
 [19, 3.5725490196078433],
 [20, 4.083333333333333],
 [21, 2.909090909090909],
 [22, 3.0673400673400675],
 [23, 3.3157894736842106],
 [24, 3.948529411764706],
 [25, 3.7411764705882353],
 [26, 2.96],
 [27, 4.171428571428572],
 [28, 3.7570093457943927],
 [29, 3.5833333333333335],
 [30, 3.488372093023256],
 [31, 3.73109243697479],
 [32, 3.625],
 [33, 3.498721227621483],
 [34, 3.8658536585365852],
 [35, 3.54040404040404],
 [36, 4.199430199430199],
 [37, 3.69811320754717],
 [38, 3.58],
 [39, 3.564516129032258],
 [40, 3.4479166666666665],
 [41, 

#### 위에서 구한 각 사용자별 평점 평균이 4점 이상인 사용자만 구해보세요~!
- 사용자 번호와 명수를 구해봅시다.(리스트는 불리언 인덱싱 사용 불가)

In [98]:
#불리언 인덱싱 적용을 위해 리스트를 배열로 전환
user_mean_arr = np.array(user_mean_list, dtype='int64')
user_mean_arr

array([[   1,    4],
       [   2,    3],
       [   3,    3],
       ...,
       [6038,    3],
       [6039,    3],
       [6040,    3]], dtype=int64)

In [99]:
b = user_mean_arr[ user_mean_arr[:, 1]>=4 ]
len(b)

1544

- 사용자가 매긴 평점의 평균이 4점 이상인 사람들은 총 1544명

#### 10번 영화의 평점 구하기!

In [100]:
movie[movie[:,1]==10][:,2].mean()

3.5405405405405403

#### 각 영화가 받은 평점의 평균 구하기!

In [101]:
movie_mean_list=[]
for i in np.unique(movie[:,1]):
    movie_mean=movie[movie[:,1]==i][:,2].mean()
    movie_mean_list.append([i,movie_mean])
movie_mean_list

[[1, 4.146846413095811],
 [2, 3.20114122681883],
 [3, 3.01673640167364],
 [4, 2.7294117647058824],
 [5, 3.0067567567567566],
 [6, 3.8787234042553194],
 [7, 3.410480349344978],
 [8, 3.014705882352941],
 [9, 2.656862745098039],
 [10, 3.5405405405405403],
 [11, 3.7938044530493706],
 [12, 2.3625],
 [13, 3.2626262626262625],
 [14, 3.542483660130719],
 [15, 2.458904109589041],
 [16, 3.7932551319648096],
 [17, 4.027544910179641],
 [18, 3.337579617834395],
 [19, 2.480719794344473],
 [20, 2.5375],
 [21, 3.6238938053097347],
 [22, 3.3492063492063493],
 [23, 2.857142857142857],
 [24, 3.1794871794871793],
 [25, 3.6510204081632653],
 [26, 3.53],
 [27, 2.9344262295081966],
 [28, 4.055865921787709],
 [29, 4.062034739454094],
 [30, 3.6486486486486487],
 [31, 3.1134751773049647],
 [32, 3.945731303772336],
 [33, 3.0],
 [34, 3.8914905768132497],
 [35, 3.3142857142857145],
 [36, 3.9579741379310347],
 [37, 3.5],
 [38, 2.8214285714285716],
 [39, 3.6233480176211454],
 [40, 3.933333333333333],
 [41, 3.9586776

In [102]:
np.unique(movie[:,1]).size   arrName[ture, false]

3706

In [113]:
ab98 = movie[[movie[:,3]>=980000000]
ab98

array([[       10,       595,         5, 980638533],
       [       10,      3447,         5, 980638373],
       [       10,      1283,         3, 980638614],
       ...,
       [     6040,       161,         3, 997454486],
       [     6040,      2725,         4, 997454180],
       [     6040,      1784,         3, 997454464]], dtype=int64)

In [128]:
bl96 = movie[movie[:,3]<960000000]
bl96

array([4, 4, 3, ..., 5, 4, 4], dtype=int64)

In [118]:
if ab98[:,2].mean() > bl96[:,2].mean():
    print("9억8천 이상 타임 평균이 더 높습니다.")
elif ab98[:,2].mean() < bl96[:,2].mean():
    print("9월 6천 미만 타임 평균이 더 높습니다")
else:
    print("두 타임의 평균이 같습니다")

9월 6천 미만 타임 평균이 더 높습니다


In [127]:
arr_N = []
arrName = np.array(arr_N)
arrName[movie[:,3]>=980000000]

IndexError: boolean index did not match indexed array along dimension 0; dimension is 0 but corresponding boolean dimension is 1000209

In [153]:
num1 = np.array([[1,2,3,4],
                 [5,6,7,8]])
num2 = np.array([[False],
               [True]])

In [154]:
num1*num2

array([[0, 0, 0, 0],
       [5, 6, 7, 8]])

### 